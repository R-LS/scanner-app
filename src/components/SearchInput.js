import {
	Button,
	Card,
	CardContent,
	FormControl,
	Grid,
	InputAdornment,
	InputLabel,
	MenuItem,
	Select,
	TextField,
	FormHelperText,
	Alert,
	LinearProgress,
	Box,
} from "@mui/material";
import React, { useEffect, useState } from "react";
import { Controller, useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";
import { useSubmitInput } from "../service/data";
import { useNavigate } from "react-router-dom";
import axios from "axios";

const schema = yup
	.object({
		projectName: yup
			.string()
			.min(4, "Must be more than 3 characters")
			.required("This field is required"),
		scanningMode: yup.string().required("This field is required"),
		scanDimensionsX: yup
			.number()
			.min(1, "Must be 1 or larger")
			.integer("Must be an integer")
			.typeError("Must be a number")
			.required("This field is required"),
		scanDimensionsY: yup
			.number()
			.min(1, "Must be 1 or larger")
			.integer("Must be an integer")
			.typeError("Must be a number")
			.required("This field is required"),
		scannerFrequency: yup
			.number()
			.min(1, "Must be 1 or larger")
			.test("is-decimal",
			"Must be maximum 1 decimal point value",
			(val) => {
			  if (val !== undefined) {
				return /^(\d)*(\.)?([0-9]{1})?$/.test(val);
			  }
			  return true;})
			.typeError("Must be a number or decimal")
			.required("This field is required")
			,
	})
	.required();

const SearchInput = () => {
	const { mutate } = useSubmitInput();
	const [errorStatus, setErrorStatus] = useState();
	const [isSending, setIsSending] = useState(false);
	
	const navigate = useNavigate();
	const {
		handleSubmit,
		control,
		formState: { errors },
	} = useForm({
		resolver: yupResolver(schema),
	});
	const onSubmit = (data) => {
		setIsSending(true);
		//console.log(data);

		//console.log("form errors:", errors);

		mutate(data, {
			onSuccess: (res) => {
				axios.get("https://wavescan-internship.saurabhmudgal.repl.co/success").then((resp) => {
					var scData = resp.data.map((row,i)=>{
						row.id=i
						return row
					})
					console.log(resp.data)
					navigate("/result",{state: {scannerData: scData}});
					
				})
				setIsSending(false);
			},
			onError: (error) => {
				setIsSending(false);
				setErrorStatus(error.response.data);
			},
		});
	};
	//console.log("form errors:", errors.projectName);

	const scanOpts = ["Gantry", "Crawler", "Auto", "Manual", "Arm"];

	useEffect(()=>{
		window.history.replaceState({}, document.title)
	})

	return (
		<Card>
			<CardContent>
				<form onSubmit={handleSubmit(onSubmit)}>
					<Grid container spacing={3}>
						<Grid item md={12} xs={12}>
							<Controller
								name={"projectName"}
								control={control}
								rules={{ required: true }}
								render={({ field: { onChange, value="" } }) => (
									<TextField
										fullWidth
										data-testid="tf-projectName"
										id="projectName"
										error={errors.projectName!==undefined}
										onChange={onChange}
										value={value}
										label={"Project Name"}
										helperText={
											errors.projectName &&
											errors.projectName.message
										}
									/>
								)}
							/>
						</Grid>
						<Grid item md={12} xs={12}>
							<FormControl fullWidth error={errors.scanningMode!==undefined}>
								<InputLabel id={"scanningMode-select-label"}>
									Scanning Mode
								</InputLabel>
								<Controller
									name={"scanningMode"}
									control={control}
									rules={{ required: true }}
									render={({
										field: { onChange, value="" },
									}) => (
										<Select
											id="scanningMode"
											data-testid ="scanningMode"
											labelId="scanningMode-label"
											value={value}
											defaultValue={scanOpts[0].toUpperCase()}
											label="Scanner Mode"
											onChange={onChange}
										>
											{scanOpts.map((scOpt, i) => (
												<MenuItem
													key={i}
													value={scOpt.toUpperCase()}
												>
													{scOpt}
												</MenuItem>
											))}
										</Select>
									)}
								/>
								{errors.scanningMode && (
									<FormHelperText>
										{errors.scanningMode.message}
									</FormHelperText>
								)}
							</FormControl>
						</Grid>
						<Grid item md={6} xs={12}>
							<Controller
								name={"scanDimensionsX"}
								control={control}
								rules={{ required: true }}
								render={({ field: { onChange, value="" } }) => (
									<TextField
										label="Scan DimensionX(cm)"
										id="scanDimensionsX"
										fullWidth
										error={errors.scanDimensionsX!==undefined}
										helperText={
											errors.scanDimensionsX &&
											errors.scanDimensionsX.message
										}
										onChange={onChange}
										value={value}
										InputProps={{
											startAdornment: (
												<InputAdornment position="start">
													x
												</InputAdornment>
											),
										}}
									/>
								)}
							/>
						</Grid>
						<Grid item md={6} xs={12}>
							<Controller
								name={"scanDimensionsY"}
								control={control}
								rules={{ required: true }}
								render={({ field: { onChange, value="" } }) => (
									<TextField
										label="Scan DimensionY(cm)"
										id="scanDimensionsY"
										fullWidth
										error={errors.scanDimensionsY!==undefined}
										helperText={
											errors.scanDimensionsY &&
											errors.scanDimensionsY.message
										}
										onChange={onChange}
										value={value}
										InputProps={{
											startAdornment: (
												<InputAdornment position="start">
													y
												</InputAdornment>
											),
										}}
									/>
								)}
							/>
						</Grid>
						<Grid item md={12} xs={12}>
							<Controller
								name={"scannerFrequency"}
								control={control}
								rules={{ required: true }}
								render={({ field: { onChange, value="" } }) => (
									<TextField
										fullWidth
										error={errors.scannerFrequency!==undefined}
										onChange={onChange}
										id="scannerFrequency"
										value={value}
										label={"Scanner Frequency (GHz)"}
										helperText={
											errors.scannerFrequency &&
											errors.scannerFrequency.message
										}
									/>
								)}
							/>
						</Grid>

						<Grid item container justifyContent="space-between">
							<Grid item xs={9} md={4}>

								<Button
									data-testid="submitButton"
									fullWidth
									variant="contained"
									type="submit"
								>
									Scan
								</Button>
							</Grid>
							{errorStatus && (
								<Grid item xs={12} md={6}>
									<Alert severity="error">
										{errorStatus}
									</Alert>
								</Grid>
							)}
						</Grid>
						{isSending && (
							<Grid item md={12}>
								<Box sx={{ width: "100%" }} data-testid="loading">
									<LinearProgress data-testid="loadingBar"/>
								</Box>
							</Grid>
						)}
					</Grid>
				</form>
			</CardContent>
		</Card>
	);
};
export default SearchInput;
