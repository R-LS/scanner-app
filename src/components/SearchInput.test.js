const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));


import SearchInput from "./SearchInput"
import { render,fireEvent,screen } from "@testing-library/react"
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import {QueryClient,QueryClientProvider}  from 'react-query';


const queryClient = new QueryClient()

beforeEach(() => {
  render(<QueryClientProvider client={queryClient}><SearchInput/></QueryClientProvider>)
});

const testInput = async(name, value, fieldError, id) => {
  const textfield = screen.getByRole("textbox",{name: name})
  fireEvent.change(textfield, {target: {value: value}})
  userEvent.click(screen.getByText('Scan'))
  var fieldError = await screen.findByText(fieldError)
  expect(fieldError).toBeTruthy()
  expect(fieldError.id).toBe(id)
}

//error message testing
describe("Input validation negative test cases",()=>{
test("Project Name must be more than 3 characters",async()=>{
  testInput("Project Name","A","Must be more than 3 characters","projectName-helper-text")
})

test("X Dimension must be 1 or greater",async()=>{
  testInput("Scan DimensionX(cm)","0","Must be 1 or larger","scanDimensionsX-helper-text")
})

test("X Dimension must be number",async()=>{
  testInput("Scan DimensionX(cm)","A","Must be a number","scanDimensionsX-helper-text")
})

test("X Dimension must be integer",async()=>{
  testInput("Scan DimensionX(cm)","1.0","Must be an integer","scanDimensionsX-helper-text")
})

test("Y Dimension must be 1 or greater",async()=>{
  testInput("Scan DimensionY(cm)","0","Must be 1 or larger","scanDimensionsY-helper-text")
})

test("Y Dimension must be number",async()=>{
  testInput("Scan DimensionY(cm)","A","Must be a number","scanDimensionsY-helper-text")
})

test("Y Dimension must be integer",async()=>{
  testInput("Scan DimensionY(cm)","1.0","Must be an integer","scanDimensionsY-helper-text")
})

test("Scanner frequency must be 1 or greater", async()=>{
    testInput("Scanner Frequency (GHz)",'0','Must be 1 or larger',"scannerFrequency-helper-text")
})

test("Scanner frequency must be number or decimal",async()=>{
  testInput("Scanner Frequency (GHz)","A","Must be a number or decimal","scannerFrequency-helper-text")
})

test("Scanner frequency must be number or decimal",async()=>{
  testInput("Scanner Frequency (GHz)","1.11","Must be maximum 1 decimal point value","scannerFrequency-helper-text")
})

test("All 5 fields should be required",async ()=>{
    // ACT
    userEvent.click(screen.getByText('Scan'))
    var fieldRequired = await screen.findAllByText('This field is required')
    //ASSERT
    expect(fieldRequired.length).toBe(5)
    expect(fieldRequired[0].id).toBe("projectName-helper-text")
    //expect(fieldRequired[1].id).toBe("scanningMode-helper-text")
    expect(fieldRequired[2].id).toBe("scanDimensionsX-helper-text")
    expect(fieldRequired[3].id).toBe("scanDimensionsY-helper-text")
    expect(fieldRequired[4].id).toBe("scannerFrequency-helper-text")
})})


