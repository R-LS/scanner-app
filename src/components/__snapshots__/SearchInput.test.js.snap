// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should be rendered correctly Snapshot testing renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="tf-projectName"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="projectName"
              id="projectName-label"
            >
              Project Name
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="projectName"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Project Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              id="scanningMode-select-label"
            >
              Scanning Mode
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-testid="scanningMode"
              onClick={[Function]}
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="scanningMode-label scanningMode"
                className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="scanningMode"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  className="notranslate"
                >
                  â€‹
                </span>
              </div>
              <input
                aria-hidden={true}
                className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                disabled={false}
                onAnimationStart={[Function]}
                onChange={[Function]}
                required={false}
                tabIndex={-1}
                value=""
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Scanner Mode
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={true}
              htmlFor="scanDimensionsX"
              id="scanDimensionsX-label"
            >
              Scan DimensionX(cm)
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <div
                className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                >
                  x
                </p>
              </div>
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="scanDimensionsX"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-14lo706"
                >
                  <span>
                    Scan DimensionX(cm)
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={true}
              htmlFor="scanDimensionsY"
              id="scanDimensionsY-label"
            >
              Scan DimensionY(cm)
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <div
                className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                >
                  y
                </p>
              </div>
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="scanDimensionsY"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-14lo706"
                >
                  <span>
                    Scan DimensionY(cm)
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="scannerFrequency"
              id="scannerFrequency-label"
            >
              Scanner Frequency (GHz)
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="scannerFrequency"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Scanner Frequency (GHz)
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-item css-9pwih8-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9 MuiGrid-grid-md-4 css-1mfid5f-MuiGrid-root"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              data-testid="submitButton"
              disabled={false}
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="submit"
            >
              Scan
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
